// @sfxcode/formkit-primevue/dist/formkit-primevue.es.js
import { markRaw as pe, defineComponent as b, ref as de, computed as h, resolveComponent as y, openBlock as f, createElementBlock as c, createVNode as _, unref as n, toDisplayString as k, createCommentVNode as w, normalizeStyle as C, normalizeClass as g, createElementVNode as X, Fragment as fe, renderList as ce } from "vue";
const me = [
  "__key",
  "__init",
  "__shim",
  "__original",
  "__index",
  "__prevKey"
];
function G() {
  return Math.random().toString(36).substring(2, 15);
}
function P(a, l) {
  return Object.prototype.hasOwnProperty.call(a, l);
}
function M(a, l, e = !0, t = ["__key"]) {
  if (a === l)
    return !0;
  if (typeof l == "object" && typeof a == "object") {
    if (a instanceof Map || a instanceof Set)
      return !1;
    if (a instanceof Date && l instanceof Date)
      return a.getTime() === l.getTime();
    if (a instanceof RegExp && l instanceof RegExp)
      return be(a, l);
    if (a === null || l === null || Object.keys(a).length !== Object.keys(l).length)
      return !1;
    for (const i of t)
      if ((i in a || i in l) && a[i] !== l[i])
        return !1;
    for (const i in a)
      if (!(i in l) || a[i] !== l[i] && !e || e && !M(a[i], l[i], e, t))
        return !1;
    return !0;
  }
  return !1;
}
function be(a, l) {
  return a.source === l.source && a.flags.split("").sort().join("") === l.flags.split("").sort().join("");
}
function j(a) {
  const l = typeof a;
  if (l === "number")
    return !1;
  if (a === void 0)
    return !0;
  if (l === "string")
    return a === "";
  if (l === "object") {
    if (a === null)
      return !0;
    for (const e in a)
      return !1;
    return !(a instanceof RegExp || a instanceof Date);
  }
  return !1;
}
function T(a) {
  return Object.prototype.toString.call(a) === "[object Object]";
}
function S(a) {
  return T(a) || Array.isArray(a);
}
function Z(a) {
  if (T(a) === !1 || a.__FKNode__ || a.__POJO__ === !1)
    return !1;
  const l = a.constructor;
  if (l === void 0)
    return !0;
  const e = l.prototype;
  return !(T(e) === !1 || e.hasOwnProperty("isPrototypeOf") === !1);
}
// @__NO_SIDE_EFFECTS__
function ee(a, l, e = !1, t = !1) {
  if (l === null)
    return null;
  const i = {};
  if (typeof l == "string")
    return l;
  for (const s in a)
    if (P(l, s) && (l[s] !== void 0 || !t)) {
      if (e && Array.isArray(a[s]) && Array.isArray(l[s])) {
        i[s] = a[s].concat(l[s]);
        continue;
      }
      if (l[s] === void 0)
        continue;
      Z(a[s]) && Z(l[s]) ? i[s] = /* @__PURE__ */ ee(a[s], l[s], e, t) : i[s] = l[s];
    } else
      i[s] = a[s];
  for (const s in l)
    !P(i, s) && l[s] !== void 0 && (i[s] = l[s]);
  return i;
}
function ye(a) {
  return a.replace(/-([a-z0-9])/gi, (l, e) => e.toUpperCase());
}
function _e(a) {
  return a.replace(/([a-z0-9])([A-Z])/g, (l, e, t) => e + "-" + t.toLowerCase()).replace(" ", "-").toLowerCase();
}
function D(a, l = me) {
  if (a === null || a instanceof RegExp || a instanceof Date || a instanceof Map || a instanceof Set || typeof File == "function" && a instanceof File)
    return a;
  let e;
  Array.isArray(a) ? e = a.map((t) => typeof t == "object" ? D(t, l) : t) : e = Object.keys(a).reduce((t, i) => (t[i] = typeof a[i] == "object" ? D(a[i], l) : a[i], t), {});
  for (const t of l)
    t in a && Object.defineProperty(e, t, {
      enumerable: !1,
      value: a[t]
    });
  return e;
}
function E(a) {
  return typeof a == "object" ? D(a) : a;
}
function H(a) {
  return a !== void 0 && a !== "false" && a !== !1 ? !0 : void 0;
}
function V(a) {
  return Object.isFrozen(a) ? a : Object.defineProperty(a, "__init", {
    enumerable: !1,
    value: !0
  });
}
function ve(a) {
  return a.normalize("NFD").replace(/[\u0300-\u036f]/g, "").toLowerCase().replace(/[^a-z0-9]/g, " ").trim().replace(/\s+/g, "-");
}
function te() {
  const a = [];
  let l = 0;
  const e = (i) => a.push(i), t = (i) => {
    const s = a[l];
    return typeof s == "function" ? s(i, (o) => (l++, t(o))) : (l = 0, i);
  };
  return e.dispatch = t, e.unshift = (i) => a.unshift(i), e.remove = (i) => {
    const s = a.indexOf(i);
    s > -1 && a.splice(s, 1);
  }, e;
}
function he() {
  const a = /* @__PURE__ */ new Map(), l = /* @__PURE__ */ new Map();
  let e;
  const t = (i, s) => {
    if (e) {
      e.set(s.name, [i, s]);
      return;
    }
    a.has(s.name) && a.get(s.name).forEach((o) => {
      (s.origin === i || o.modifiers.includes("deep")) && o.listener(s);
    }), s.bubble && i.bubble(s);
  };
  return t.flush = () => {
    a.clear(), l.clear(), e == null || e.clear();
  }, t.on = (i, s) => {
    const [o, ...r] = i.split("."), u = s.receipt || G(), p = {
      modifiers: r,
      event: o,
      listener: s,
      receipt: u
    };
    return a.has(o) ? a.get(o).push(p) : a.set(o, [p]), l.has(u) ? l.get(u).push(o) : l.set(u, [o]), u;
  }, t.off = (i) => {
    var s;
    l.has(i) && ((s = l.get(i)) === null || s === void 0 || s.forEach((o) => {
      const r = a.get(o);
      Array.isArray(r) && a.set(o, r.filter((u) => u.receipt !== i));
    }), l.delete(i));
  }, t.pause = (i) => {
    e || (e = /* @__PURE__ */ new Map()), i && i.walk((s) => s._e.pause());
  }, t.play = (i) => {
    if (!e)
      return;
    const s = e;
    e = void 0, s.forEach(([o, r]) => t(o, r)), i && i.walk((o) => o._e.play());
  }, t;
}
function ge(a, l, e, t, i = !0, s) {
  return l._e(a, {
    payload: t,
    name: e,
    bubble: i,
    origin: a,
    meta: s
  }), a;
}
function $e(a, l, e) {
  return K(a.parent) && a.parent._e(a.parent, e), a;
}
function we(a, l, e, t) {
  return l._e.on(e, t);
}
function ke(a, l, e) {
  return l._e.off(e), a;
}
const U = te();
U((a, l) => (a.message || (a.message = `E${a.code}`), l(a)));
const N = te();
N((a, l) => {
  a.message || (a.message = `W${a.code}`);
  const e = l(a);
  return console && typeof console.warn == "function" && console.warn(e.message), e;
});
function ae(a, l = {}) {
  N.dispatch({ code: a, data: l });
}
function L(a, l = {}) {
  throw Error(U.dispatch({ code: a, data: l }).message);
}
function B(a, l) {
  const e = {
    blocking: !1,
    key: G(),
    meta: {},
    type: "state",
    visible: !0,
    ...a
  };
  return l && e.value && e.meta.localize !== !1 && (e.value = l.t(e), e.meta.locale = l.config.locale), e;
}
function Ce(a, ...l) {
  const e = `${a.name}-set`, t = (i) => B({
    key: ve(i),
    type: "error",
    value: i,
    meta: { source: e, autoClear: !0 }
  });
  return l.filter((i) => !!i).map((i) => {
    if (typeof i == "string" && (i = [i]), Array.isArray(i))
      return i.map((s) => t(s));
    {
      const s = {};
      for (const o in i)
        Array.isArray(i[o]) ? s[o] = i[o].map((r) => t(r)) : s[o] = [t(i[o])];
      return s;
    }
  });
}
const ne = /* @__PURE__ */ new Map(), A = /* @__PURE__ */ new Map(), Ve = he();
function Le(a) {
  if (A.has(a)) {
    const l = A.get(a);
    A.delete(a), ne.delete(l), Ve(a, {
      payload: null,
      name: l,
      bubble: !1,
      origin: a
    });
  }
}
function Ie(a) {
  return ne.get(a);
}
function W(a, l, e) {
  let t = !0;
  return l in a.config._t ? t = !1 : a.emit(`config:${l}`, e, !1), l in a.props || (a.emit("prop", { prop: l, value: e }), a.emit(`prop:${l}`, e)), t;
}
function Se(a) {
  const l = document.getElementById(a);
  if (l instanceof HTMLFormElement) {
    const e = new Event("submit", { cancelable: !0, bubbles: !0 });
    l.dispatchEvent(e);
    return;
  }
  ae(151, a);
}
function Oe(a) {
  const l = (e) => {
    for (const t in e.store) {
      const i = e.store[t];
      i.type === "error" || i.type === "ui" && t === "incomplete" ? e.store.remove(t) : i.type === "state" && e.store.set({ ...i, value: !1 });
    }
  };
  l(a), a.walk(l);
}
function Pe(a, l) {
  const e = typeof a == "string" ? Ie(a) : a;
  if (e) {
    const t = (o) => E(o.props.initial) || (o.type === "group" ? {} : o.type === "list" ? [] : void 0);
    e._e.pause(e);
    const i = E(l);
    return l && !j(l) && (e.props.initial = S(i) ? V(i) : i, e.props._init = e.props.initial), e.input(t(e), !1), e.walk((o) => {
      o.type === "list" && o.sync || o.input(t(o), !1);
    }), e.input(j(i) && i ? i : t(e), !1), e.type !== "input" && l && !j(l) && S(l) && e.walk((o) => {
      o.props.initial = S(o.value) ? V(o.value) : o.value, o.props._init = e.props.initial;
    }), e._e.play(e), Oe(e), e.emit("reset", e), e;
  }
  ae(152, a);
}
const De = {
  delimiter: ".",
  delay: 0,
  locale: "en",
  rootClasses: (a) => ({ [`formkit-${_e(a)}`]: !0 })
}, Ee = Symbol("index"), F = Symbol("removed"), z = Symbol("moved"), le = Symbol("inserted");
function Re(a) {
  return a.type === "list" && Array.isArray(a._value);
}
function K(a) {
  return a && typeof a == "object" && a.__FKNode__ === !0;
}
const O = (a, l, e) => {
  L(102, [a, e]);
};
d(Ge, O, !1), d(Ze), d(xe), d(tt, O, !1), d(at), d($e), d(dt), d(Ne), d(!1), d(Ke), d(Ue), d(Be), d(ft), d(Fe), d(Xe, Qe, !1), d(Ae), d(Ye), d(ge), d(lt), d(we), d(ke), d(!1, He), d(!1), d(We), d(st, O, !1), d(pt), d(Je), d(re), d(ut), d(ot), d(se), d(et, !1, !1), d(qe);
function d(a, l, e = !0) {
  return {
    get: a ? (t, i) => e ? (...s) => a(t, i, ...s) : a(t, i) : !1,
    set: l !== void 0 ? l : O.bind(null)
  };
}
let Me = 0;
function je(a) {
  return a.type === "group" ? V(a.value && typeof a.value == "object" && !Array.isArray(a.value) ? a.value : {}) : a.type === "list" ? V(Array.isArray(a.value) ? a.value : []) : a.value;
}
function Ae(a, l, e, t = !0) {
  return l._value = Te(a, a.hook.input.dispatch(e)), a.emit("input", l._value), a.isCreated && a.type === "input" && M(l._value, l.value) ? (a.emit("commitRaw", l.value), l.settled) : (l.isSettled && a.disturb(), t ? (l._tmo && clearTimeout(l._tmo), l._tmo = setTimeout(R, a.props.delay, a, l)) : R(a, l), l.settled);
}
function Te(a, l) {
  switch (a.type) {
    case "input":
      break;
    case "group":
      (!l || typeof l != "object") && L(107, [a, l]);
      break;
    case "list":
      Array.isArray(l) || L(108, [a, l]);
      break;
  }
  return l;
}
function R(a, l, e = !0, t = !0) {
  l._value = l.value = a.hook.commit.dispatch(l._value), a.type !== "input" && t && a.hydrate(), a.emit("commitRaw", l.value), a.emit("commit", l.value), e && a.calm();
}
function ie(a, { name: l, value: e, from: t }) {
  if (!Object.isFrozen(a._value)) {
    if (Re(a)) {
      const i = e === F ? [] : e === z && typeof t == "number" ? a._value.splice(t, 1) : [e];
      a._value.splice(l, e === z || t === le ? 0 : 1, ...i);
      return;
    }
    e !== F ? a._value[l] = e : delete a._value[l];
  }
}
function Fe(a, l) {
  const e = l._value;
  return a.type === "list" && a.sync && ze(a, l), l.children.forEach((t) => {
    if (typeof e == "object")
      if (t.name in e) {
        const i = t.type !== "input" || e[t.name] && typeof e[t.name] == "object" ? V(e[t.name]) : e[t.name];
        if (!t.isSettled || !S(i) && M(i, t._value))
          return;
        t.input(i, !1);
      } else
        (a.type !== "list" || typeof t.name == "number") && ie(l, { name: t.name, value: t.value }), e.__init || (t.type === "group" ? t.input({}, !1) : t.type === "list" ? t.input([], !1) : t.input(void 0, !1));
  }), a;
}
function ze(a, l) {
  const e = a._value;
  if (!Array.isArray(e))
    return;
  const t = [], i = new Set(l.children), s = /* @__PURE__ */ new Map();
  e.forEach((r, u) => {
    if (l.children[u] && l.children[u]._value === r)
      t.push(l.children[u]), i.delete(l.children[u]);
    else {
      t.push(null);
      const p = s.get(r) || [];
      p.push(u), s.set(r, p);
    }
  }), i.size && s.size && i.forEach((r) => {
    if (s.has(r._value)) {
      const u = s.get(r._value), p = u.shift();
      t[p] = r, i.delete(r), u.length || s.delete(r._value);
    }
  });
  const o = [];
  for (s.forEach((r) => {
    o.push(...r);
  }); i.size && o.length; ) {
    const r = i.values().next().value, u = o.shift();
    if (u === void 0)
      break;
    t[u] = r, i.delete(r);
  }
  o.forEach((r, u) => {
    t[r] = ct({ value: u });
  }), i.size && i.forEach((r) => {
    if (!("__FKP" in r)) {
      const u = r._c.parent;
      if (!u || mt(u))
        return;
      u.ledger.unmerge(r), r._c.parent = null, r.destroy();
    }
  }), l.children = t;
}
function Ue(a, l) {
  var e;
  return l._d <= 0 && (l.isSettled = !1, a.emit("settled", !1, !1), l.settled = new Promise((t) => {
    l._resolve = t;
  }), a.parent && ((e = a.parent) === null || e === void 0 || e.disturb())), l._d++, a;
}
function Ne(a, l, e) {
  var t;
  if (e !== void 0 && a.type !== "input")
    return ie(l, e), R(a, l, !0, !1);
  l._d > 0 && l._d--, l._d === 0 && (l.isSettled = !0, a.emit("settled", !0, !1), a.parent && ((t = a.parent) === null || t === void 0 || t.calm({ name: a.name, value: l.value })), l._resolve && l._resolve(l.value));
}
function Be(a, l) {
  a.emit("destroying", a), a.store.filter(() => !1), a.parent && (a.parent.emit("childRemoved", a), a.parent.remove(a)), Le(a), a.emit("destroyed", a), l._e.flush(), l._value = l.value = void 0;
  for (const e in l.context)
    delete l.context[e];
  l.plugins.clear(), l.context = null;
}
function Ke(a, l, e) {
  l.type = e.type, l.props.definition = D(e), l.value = l._value = je({
    type: a.type,
    value: l.value
  }), e.forceTypeProp && (a.props.type && (a.props.originalType = a.props.type), l.props.type = e.forceTypeProp), e.family && (l.props.family = e.family), e.features && e.features.forEach((t) => t(a)), e.props && a.addProps(e.props), a.emit("defined", e);
}
function xe(a, l, e) {
  var t;
  if (a.props.attrs) {
    const i = { ...a.props.attrs };
    a.props._emit = !1;
    for (const o in i) {
      const r = ye(o);
      e.includes(r) && (a.props[r] = i[o], delete i[o]);
    }
    const s = E(l._value);
    a.props.initial = a.type !== "input" ? V(s) : s, a.props._emit = !0, a.props.attrs = i, a.props.definition && (a.props.definition.props = [
      ...((t = a.props.definition) === null || t === void 0 ? void 0 : t.props) || [],
      ...e
    ]);
  }
  return a.emit("added-props", e), a;
}
function Ze(a, l, e, t) {
  if (a.type === "input" && L(100, a), e.parent && e.parent !== a && e.parent.remove(e), !l.children.includes(e)) {
    if (t !== void 0 && a.type === "list") {
      const i = l.children[t];
      i && "__FKP" in i ? (e._c.uid = i.uid, l.children.splice(t, 1, e)) : l.children.splice(t, 0, e), Array.isArray(a.value) && a.value.length < l.children.length && a.disturb().calm({
        name: t,
        value: e.value,
        from: le
      });
    } else
      l.children.push(e);
    e.isSettled || a.disturb();
  }
  if (e.parent !== a) {
    if (e.parent = a, e.parent !== a)
      return a.remove(e), e.parent.add(e), a;
  } else
    e.use(a.plugins);
  return R(a, l, !1), a.ledger.merge(e), a.emit("child", e), a;
}
function He(a, l, e, t) {
  return K(t) ? (a.parent && a.parent !== t && a.parent.remove(a), l.parent = t, a.resetConfig(), t.children.includes(a) ? a.use(t.plugins) : t.add(a), !0) : t === null ? (l.parent = null, !0) : !1;
}
function We(a, l, e) {
  const t = l.children.indexOf(e);
  if (t !== -1) {
    e.isSettled && a.disturb(), l.children.splice(t, 1);
    let i = H(e.props.preserve), s = e.parent;
    for (; i === void 0 && s; )
      i = H(s.props.preserve), s = s.parent;
    i ? a.calm() : a.calm({
      name: a.type === "list" ? t : e.name,
      value: F
    }), e.parent = null, e.config._rmn = e;
  }
  return a.ledger.unmerge(e), a;
}
function Ye(a, l, e) {
  l.children.forEach((t) => !("__FKP" in t) && e(t));
}
function qe(a, l, e, t = !1, i = !1) {
  l.children.some((s) => {
    if ("__FKP" in s)
      return !1;
    const o = e(s);
    return t && o === !1 ? !0 : i && o === !1 ? !1 : s.walk(e, t, i);
  });
}
function Je(a, l) {
  const e = a.parent || void 0;
  l.config = rt(a.config._t, e), a.walk((t) => t.resetConfig());
}
function se(a, l, e, t = !0, i = !0) {
  return Array.isArray(e) || e instanceof Set ? (e.forEach((s) => se(a, l, s)), a) : (l.plugins.has(e) || (i && typeof e.library == "function" && e.library(a), t && e(a) !== !1 && (l.plugins.add(e), a.children.forEach((s) => s.use(e)))), a);
}
function Qe(a, l, e, t) {
  if (K(a.parent)) {
    const i = a.parent.children, s = t >= i.length ? i.length - 1 : t < 0 ? 0 : t, o = i.indexOf(a);
    return o === -1 ? !1 : (i.splice(o, 1), i.splice(s, 0, a), a.parent.children = i, a.parent.type === "list" && a.parent.disturb().calm({ name: s, value: z, from: o }), !0);
  }
  return !1;
}
function Xe(a) {
  if (a.parent) {
    const l = [...a.parent.children].indexOf(a);
    return l === -1 ? a.parent.children.length : l;
  }
  return -1;
}
function Ge(a, l) {
  return l;
}
function et(a, l) {
  var e;
  return ((e = a.parent) === null || e === void 0 ? void 0 : e.type) === "list" ? a.index : l.name !== Ee ? l.name : a.index;
}
function tt(a, l) {
  return l.parent ? l.parent.address.concat([a.name]) : [a.name];
}
function at(a, l, e) {
  const t = typeof e == "string" ? e.split(a.config.delimiter) : e;
  if (!t.length)
    return;
  const i = t[0];
  let s = a.parent;
  for (s || (String(t[0]) === String(a.name) && t.shift(), s = a), i === "$parent" && t.shift(); s && t.length; ) {
    const o = t.shift();
    switch (o) {
      case "$root":
        s = a.root;
        break;
      case "$parent":
        s = s.parent;
        break;
      case "$self":
        s = a;
        break;
      default:
        s = s.children.find((r) => !("__FKP" in r) && String(r.name) === String(o)) || nt(s, o);
    }
  }
  return s || void 0;
}
function nt(a, l) {
  const e = String(l).match(/^(find)\((.*)\)$/);
  if (e) {
    const [, t, i] = e, s = i.split(",").map((o) => o.trim());
    switch (t) {
      case "find":
        return a.find(s[0], s[1]);
      default:
        return;
    }
  }
}
function lt(a, l, e, t) {
  return it(a, e, t);
}
function it(a, l, e = "name") {
  const t = typeof e == "string" ? (s) => s[e] == l : e, i = [a];
  for (; i.length; ) {
    const s = i.shift();
    if (!("__FKP" in s)) {
      if (t(s, l))
        return s;
      i.push(...s.children);
    }
  }
}
function st(a) {
  let l = a;
  for (; l.parent; )
    l = l.parent;
  return l;
}
function rt(a = {}, l) {
  let e;
  return new Proxy(a, {
    get(...t) {
      const i = t[1];
      if (i === "_t")
        return a;
      const s = Reflect.get(...t);
      if (s !== void 0)
        return s;
      if (l) {
        const o = l.config[i];
        if (o !== void 0)
          return o;
      }
      if (a.rootConfig && typeof i == "string") {
        const o = a.rootConfig[i];
        if (o !== void 0)
          return o;
      }
      return i === "delay" && (e == null ? void 0 : e.type) === "input" ? 20 : De[i];
    },
    set(...t) {
      const i = t[1], s = t[2];
      if (i === "_n")
        return e = s, a.rootConfig && a.rootConfig._add(e), !0;
      if (i === "_rmn")
        return a.rootConfig && a.rootConfig._rm(e), e = void 0, !0;
      if (!M(a[i], s, !1)) {
        const o = Reflect.set(...t);
        return e && (e.emit(`config:${i}`, s, !1), W(e, i, s), e.walk((r) => W(r, i, s), !1, !0)), o;
      }
      return !0;
    }
  });
}
function ot(a, l, e, t = "ui") {
  const i = typeof e == "string" ? { key: e, value: e, type: t } : e, s = a.hook.text.dispatch(i);
  return a.emit("text", s, !1), s.value;
}
function ut(a) {
  const l = a.name;
  do {
    if (a.props.isForm === !0)
      break;
    a.parent || L(106, l), a = a.parent;
  } while (a);
  a.props.id && Se(a.props.id);
}
function pt(a, l, e) {
  return Pe(a, e);
}
function re(a, l, e, t) {
  const i = `${a.name}-set`, s = a.hook.setErrors.dispatch({ localErrors: e, childErrors: t });
  return Ce(a, s.localErrors, s.childErrors).forEach((o) => {
    a.store.apply(o, (r) => r.meta.source === i);
  }), a;
}
function dt(a, l, e = !0, t) {
  return re(a, l, []), e && (t = t || `${a.name}-set`, a.walk((i) => {
    i.store.filter((s) => !(s.type === "error" && s.meta && s.meta.source === t));
  })), a;
}
function ft(a, l, e, t) {
  return l.traps.set(e, t), a;
}
function ct(a) {
  var l, e, t, i;
  return {
    __FKP: !0,
    uid: Symbol(),
    name: (l = a == null ? void 0 : a.name) !== null && l !== void 0 ? l : `p_${Me++}`,
    value: (e = a == null ? void 0 : a.value) !== null && e !== void 0 ? e : null,
    _value: (t = a == null ? void 0 : a.value) !== null && t !== void 0 ? t : null,
    type: (i = a == null ? void 0 : a.type) !== null && i !== void 0 ? i : "input",
    use: () => {
    },
    input(s) {
      return this._value = s, this.value = s, Promise.resolve();
    },
    isSettled: !0
  };
}
function mt(a) {
  return "__FKP" in a;
}
function Y(a) {
  return typeof a != "string" && P(a, "$el");
}
function bt(a) {
  return typeof a != "string" && P(a, "$cmp");
}
// @__NO_SIDE_EFFECTS__
function $(a, l, e = !1) {
  return /* @__NO_SIDE_EFFECTS__ */ (...i) => {
    const s = (o) => {
      const r = !l || typeof l == "string" ? { $el: l } : l();
      return (Y(r) || bt(r)) && (r.meta || (r.meta = { section: a }), i.length && !r.children && (r.children = [
        ...i.map((u) => typeof u == "string" ? u : u(o))
      ]), Y(r) && (r.attrs = {
        class: `$classes.${a}`,
        ...r.attrs || {}
      })), {
        if: `$slots.${a}`,
        then: `$slots.${a}`,
        else: a in o ? /* @__PURE__ */ _t(r, o[a]) : r
      };
    };
    return s._s = a, e ? /* @__PURE__ */ yt(s) : s;
  };
}
// @__NO_SIDE_EFFECTS__
function yt(a) {
  return (l) => [a(l)];
}
function q(a) {
  return typeof a == "object" && ("$el" in a || "$cmp" in a || "$formkit" in a);
}
// @__NO_SIDE_EFFECTS__
function _t(a, l = {}) {
  return typeof a == "string" ? q(l) || typeof l == "string" ? l : a : Array.isArray(a) ? q(l) ? l : a : /* @__PURE__ */ ee(a, l);
}
const vt = /* @__PURE__ */ $("help", () => ({
  $el: "div",
  if: "$help",
  attrs: {
    id: '$: "help-" + $id'
  }
})), ht = /* @__PURE__ */ $("inner", "div"), gt = /* @__PURE__ */ $("label", () => ({
  $el: "label",
  if: "$label",
  attrs: {
    for: "$id"
  }
})), $t = /* @__PURE__ */ $("message", () => ({
  $el: "li",
  for: ["message", "$messages"],
  attrs: {
    key: "$message.key",
    id: "$id + '-' + $message.key",
    "data-message-type": "$message.type"
  }
})), wt = /* @__PURE__ */ $("messages", () => ({
  $el: "ul",
  if: "$defaultMessagePlacement && $fns.length($messages)"
})), kt = /* @__PURE__ */ $("outer", () => ({
  $el: "div",
  attrs: {
    key: "$id",
    "data-family": "$family || undefined",
    "data-type": "$type",
    "data-multiple": '$attrs.multiple || ($type != "select" && $options != undefined) || undefined',
    "data-disabled": '$: ($disabled !== "false" && $disabled) || undefined',
    "data-empty": "$state.empty || undefined",
    "data-complete": "$state.complete || undefined",
    "data-invalid": "$state.valid === false && $state.validationVisible || undefined",
    "data-errors": "$state.errors || undefined",
    "data-submitted": "$state.submitted || undefined",
    "data-prefix-icon": "$_rawPrefixIcon !== undefined || undefined",
    "data-suffix-icon": "$_rawSuffixIcon !== undefined || undefined",
    "data-prefix-icon-click": "$onPrefixIconClick !== undefined || undefined",
    "data-suffix-icon-click": "$onSuffixIconClick !== undefined || undefined"
  }
})), Ct = /* @__PURE__ */ $("prefix", null), Vt = /* @__PURE__ */ $("suffix", null), Lt = /* @__PURE__ */ $("wrapper", "div");
B({
  key: "loading",
  value: !0,
  visible: !1
});
// @__NO_SIDE_EFFECTS__
function It(a) {
  return /* @__PURE__ */ kt(/* @__PURE__ */ Lt(/* @__PURE__ */ gt("$label"), /* @__PURE__ */ ht(/* @__PURE__ */ Ct(), a(), /* @__PURE__ */ Vt())), /* @__PURE__ */ vt("$help"), /* @__PURE__ */ wt(/* @__PURE__ */ $t("$message.value")));
}
B({
  type: "state",
  blocking: !0,
  visible: !1,
  value: !0,
  key: "validating"
});
new Promise((a) => {
});
const St = typeof window < "u" && typeof fetch < "u";
St && getComputedStyle(document.documentElement);
let x = !1;
const J = {
  /**
   * FormKit errors:
   */
  100: ({ data: a }) => `Only groups, lists, and forms can have children (${a.name}).`,
  101: ({ data: a }) => `You cannot directly modify the store (${a.name}). See: https://formkit.com/advanced/core#message-store`,
  102: ({ data: [a, l] }) => `You cannot directly assign node.${l} (${a.name})`,
  103: ({ data: [a] }) => `Schema expressions cannot start with an operator (${a})`,
  104: ({ data: [a, l] }) => `Schema expressions cannot end with an operator (${a} in "${l}")`,
  105: ({ data: a }) => `Invalid schema expression: ${a}`,
  106: ({ data: a }) => `Cannot submit because (${a}) is not in a form.`,
  107: ({ data: [a, l] }) => `Cannot set ${a.name} to non object value: ${l}`,
  108: ({ data: [a, l] }) => `Cannot set ${a.name} to non array value: ${l}`,
  /**
   * Input specific errors:
   */
  300: ({ data: [a] }) => `Cannot set behavior prop to overscroll (on ${a.name} input) when options prop is a function.`,
  /**
   * FormKit vue errors:
   */
  600: ({ data: a }) => `Unknown input type${typeof a.props.type == "string" ? ' "' + a.props.type + '"' : ""} ("${a.name}")`,
  601: ({ data: a }) => `Input definition${typeof a.props.type == "string" ? ' "' + a.props.type + '"' : ""} is missing a schema or component property (${a.name}).`
}, Q = {
  /**
   * Core warnings:
   */
  150: ({ data: a }) => `Schema function "${a}()" is not a valid function.`,
  151: ({ data: a }) => `No form element with id: ${a}`,
  152: ({ data: a }) => `No input element with id: ${a}`,
  /**
   * Input specific warnings:
   */
  350: ({ data: { node: a, inputType: l } }) => `Invalid options prop for ${a.name} input (${l}). See https://formkit.com/inputs/${l}`,
  /**
   * Vue warnings:
   */
  650: 'Schema "$get()" must use the id of an input to access.',
  651: ({ data: a }) => `Cannot setErrors() on "${a}" because no such id exists.`,
  652: ({ data: a }) => `Cannot clearErrors() on "${a}" because no such id exists.`,
  /**
   * Deprecation warnings:
   */
  800: ({ data: a }) => `${a} is deprecated.`
}, Ot = (a, l) => {
  if (a.code in J) {
    const e = J[a.code];
    a.message = typeof e == "function" ? e(a) : e;
  }
  return l(a);
};
x || U(Ot);
const Pt = (a, l) => {
  if (a.code in Q) {
    const e = Q[a.code];
    a.message = typeof e == "function" ? e(a) : e;
  }
  return l(a);
};
x || N(Pt);
x = !0;
let Dt = 1;
function Et(a) {
  return typeof a == "function" && a.length === 2 || typeof a == "object" && !Array.isArray(a) && !("$el" in a) && !("$cmp" in a) && !("if" in a);
}
function v(a, l = {}) {
  const e = {
    type: "input",
    ...l
  };
  let t;
  if (Et(a)) {
    const i = `SchemaComponent${Dt++}`;
    t = /* @__PURE__ */ $("input", () => ({
      $cmp: i,
      props: {
        context: "$node.context"
      }
    })), e.library = { [i]: pe(a) };
  } else
    typeof a == "function" ? t = a : t = /* @__PURE__ */ $("input", () => E(a));
  return e.schema = /* @__PURE__ */ It(t || "Schema undefined"), e.schemaMemoKey || (e.schemaMemoKey = `${Math.random()}`), e;
}
const Rt = /* @__PURE__ */ b({
  __name: "PrimeAutoComplete",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs, i = de([]);
    function s(r) {
      i.value = t.complete(r.query);
    }
    function o(r) {
      var u;
      e == null || e.node.input((u = l.context) == null ? void 0 : u._value);
    }
    return h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class), (r, u) => {
      var m;
      const p = y("AutoComplete");
      return f(), c("div", null, [
        _(p, {
          id: n(e).id,
          modelValue: n(e)._value,
          "onUpdate:modelValue": u[0] || (u[0] = (I) => n(e)._value = I),
          disabled: n(t)._disabled ?? !1,
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          suggestions: n(i),
          dropdown: ((m = n(t)) == null ? void 0 : m.dropdown) ?? !1,
          multiple: n(t).multiple ?? !1,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onComplete: s,
          onChange: o
        }, null, 8, ["id", "modelValue", "disabled", "tabindex", "aria-label", "aria-labelledby", "suggestions", "dropdown", "multiple", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), Mt = /* @__PURE__ */ b({
  __name: "PrimeCalendar",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(r) {
      e == null || e.node.input(e == null ? void 0 : e._value);
    }
    function s(r) {
      e == null || e.node.input(r);
    }
    const o = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (r, u) => {
      const p = y("Calendar");
      return f(), c("div", null, [
        _(p, {
          modelValue: n(e)._value,
          "onUpdate:modelValue": u[0] || (u[0] = (m) => n(e)._value = m),
          "input-id": l.context.id,
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "input-style": n(t).style,
          "input-class": n(o),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          "date-format": n(t).dateFormat,
          "selection-mode": n(t).selectionMode ?? "single",
          inline: n(t).inline ?? !1,
          "show-other-months": n(t).showOtherMonths ?? !0,
          "select-other-months": n(t).selectOtherMonths ?? !1,
          icon: n(t).icon,
          "show-icon": n(e).showIcon,
          "previous-icon": n(t).previousIcon ?? "pi pi-chevron-left",
          "next-icon": n(t).nextIcon ?? "pi pi-chevron-right",
          "increment-icon": n(t).incrementIcon ?? "pi pi-chevron-up",
          "decrement-icon": n(t).decrementIcon ?? "pi pi-chevron-down",
          "number-of-months": n(t).numberOfMonths ?? 1,
          "responsive-options": n(t).responsiveOptions,
          view: n(t).view ?? "date",
          "touch-u-i": n(t).touchUI ?? !1,
          "min-date": n(t).minDate,
          "max-date": n(t).maxDate,
          "disabled-dates": n(t).disabledDates,
          "disabled-days": n(t).disabledDays,
          "max-date-count": n(t).maxDateCount,
          "show-on-focus": n(t).showOnFocus ?? !0,
          "auto-z-index": n(t).autoZIndex ?? !0,
          "base-z-index": n(t).baseZIndex ?? 0,
          "show-button-bar": n(t).showButtonBar ?? !1,
          "show-time": n(t).showTime ?? !1,
          "time-only": n(t).timeOnly ?? !1,
          "short-year-cutoff": n(t).shortYearCutoff ?? "+10",
          "hour-format": n(t).hourFormat ?? "24",
          "step-hour": n(t).stepHour ?? 1,
          "step-minute": n(t).stepMinute ?? 1,
          "step-second": n(t).stepSecond ?? 1,
          "show-seconds": n(t).showSeconds ?? !1,
          "hide-on-date-time-select": n(t).hideOnDateTimeSelect ?? !1,
          "hide-on-range-selection": n(t).hideOnRangeSelection ?? !1,
          "time-separator": n(t).timeSeparator ?? ":",
          "show-week": n(t).showWeek ?? !1,
          "manual-input": n(t).manualInput ?? !0,
          "append-to": n(t).appendTo ?? "body",
          "panel-style": n(t).panelStyle,
          "panel-class": n(t).panelClass,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onDateSelect: s,
          onInput: i
        }, null, 8, ["modelValue", "input-id", "disabled", "readonly", "input-style", "input-class", "tabindex", "aria-label", "aria-labelledby", "date-format", "selection-mode", "inline", "show-other-months", "select-other-months", "icon", "show-icon", "previous-icon", "next-icon", "increment-icon", "decrement-icon", "number-of-months", "responsive-options", "view", "touch-u-i", "min-date", "max-date", "disabled-dates", "disabled-days", "max-date-count", "show-on-focus", "auto-z-index", "base-z-index", "show-button-bar", "show-time", "time-only", "short-year-cutoff", "hour-format", "step-hour", "step-minute", "step-second", "show-seconds", "hide-on-date-time-select", "hide-on-range-selection", "time-separator", "show-week", "manual-input", "append-to", "panel-style", "panel-class", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), jt = {
  key: 0,
  class: "formkit-prime-left"
}, At = {
  key: 1,
  class: "formkit-prime-right"
}, Tt = /* @__PURE__ */ b({
  __name: "PrimeCheckbox",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      var r;
      e == null || e.node.input((r = l.context) == null ? void 0 : r._value);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      const u = y("Checkbox");
      return f(), c("div", null, [
        n(e).attrs.labelLeft ? (f(), c("span", jt, k(n(e).attrs.labelLeft), 1)) : w("", !0),
        _(u, {
          modelValue: n(e)._value,
          "onUpdate:modelValue": r[0] || (r[0] = (p) => n(e)._value = p),
          "input-id": n(e).id,
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "input-style": n(t).style,
          "input-class": n(s),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          binary: n(t).binary ?? !0,
          "true-value": n(t).trueValue ?? void 0,
          "false-value": n(t).falseValue ?? void 0,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onInput: i
        }, null, 8, ["modelValue", "input-id", "disabled", "readonly", "input-style", "input-class", "tabindex", "aria-label", "aria-labelledby", "binary", "true-value", "false-value", "pt", "pt-options", "unstyled"]),
        n(e).attrs.labelRight ? (f(), c("span", At, k(n(e).attrs.labelRight), 1)) : w("", !0)
      ]);
    };
  }
}), Ft = /* @__PURE__ */ b({
  __name: "PrimeChips",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      var r;
      e == null || e.node.input((r = l.context) == null ? void 0 : r._value);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      const u = y("Chips");
      return f(), c("div", null, [
        _(u, {
          modelValue: n(e)._value,
          "onUpdate:modelValue": r[0] || (r[0] = (p) => n(e)._value = p),
          "input-id": n(e).id,
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "input-style": n(t).style,
          "input-class": n(s),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          "allow-duplicate": n(t).allowDuplicate ?? !0,
          "add-on-blur": n(t).addOnBlur ?? !1,
          max: n(t).max ?? void 0,
          placeholder: n(t).placeholder,
          separator: n(t).separator,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onAdd: i,
          onRemove: i
        }, null, 8, ["modelValue", "input-id", "disabled", "readonly", "input-style", "input-class", "tabindex", "aria-label", "aria-labelledby", "allow-duplicate", "add-on-blur", "max", "placeholder", "separator", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), zt = /* @__PURE__ */ b({
  __name: "PrimeColorPicker",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(s) {
      var o;
      e == null || e.node.input((o = l.context) == null ? void 0 : o._value);
    }
    return (s, o) => {
      const r = y("ColorPicker");
      return f(), c("div", null, [
        _(r, {
          modelValue: n(e)._value,
          "onUpdate:modelValue": o[0] || (o[0] = (u) => n(e)._value = u),
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          style: C(n(t).style),
          "panel-class": n(t).class,
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          "default-color": n(t).defaultColor ?? "ff0000",
          inline: n(t).inline ?? !1,
          format: n(t).format,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onChange: i
        }, null, 8, ["modelValue", "disabled", "readonly", "style", "panel-class", "tabindex", "aria-label", "aria-labelledby", "default-color", "inline", "format", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), Ut = /* @__PURE__ */ b({
  __name: "PrimeDropdown",
  props: {
    context: Object
  },
  setup(a) {
    const e = a.context, t = e == null ? void 0 : e.attrs;
    function i(r) {
      e == null || e.handlers.blur(r.value);
    }
    function s(r) {
      e == null || e.node.input(r.value);
    }
    const o = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (r, u) => {
      const p = y("Dropdown");
      return f(), c("div", null, [
        _(p, {
          modelValue: n(e)._value,
          "onUpdate:modelValue": u[0] || (u[0] = (m) => n(e)._value = m),
          "input-id": n(e).id,
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "input-style": n(o),
          "input-class": n(t).class,
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          options: n(t).options,
          "option-label": n(t).optionLabel ?? "label",
          "option-value": n(t).optionValue ?? "value",
          placeholder: n(t).placeholder,
          filter: n(t).filter ?? !1,
          "show-clear": n(t).showClear ?? !1,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onChange: s,
          onBlur: i
        }, null, 8, ["modelValue", "input-id", "disabled", "readonly", "input-style", "input-class", "tabindex", "aria-label", "aria-labelledby", "options", "option-label", "option-value", "placeholder", "filter", "show-clear", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), Nt = /* @__PURE__ */ b({
  __name: "PrimeEditor",
  props: {
    context: Object
  },
  setup(a) {
    const e = a.context, t = e == null ? void 0 : e.attrs;
    function i(r) {
      e == null || e.node.input(r.htmlValue);
    }
    function s(r) {
      r.range === null && (e == null || e.handlers.blur(r.htmlValue));
    }
    const o = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (r, u) => {
      const p = y("Editor");
      return f(), c("div", null, [
        _(p, {
          id: n(e).id,
          modelValue: n(e)._value,
          "onUpdate:modelValue": u[0] || (u[0] = (m) => n(e)._value = m),
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "editor-style": n(t).style,
          class: g(n(o)),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          placeholder: n(t).placeholder,
          formats: n(t).formats,
          modules: n(t).modules,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onTextChange: i,
          onSelectionChange: s
        }, null, 8, ["id", "modelValue", "disabled", "readonly", "editor-style", "class", "tabindex", "aria-label", "aria-labelledby", "placeholder", "formats", "modules", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), Bt = /* @__PURE__ */ b({
  __name: "PrimeInputMask",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      var r, u;
      e == null || e.node.input((r = l.context) == null ? void 0 : r._value), e == null || e.handlers.blur((u = l.context) == null ? void 0 : u._value);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      const u = y("InputMask");
      return f(), c("div", null, [
        _(u, {
          id: n(e).id,
          modelValue: n(e)._value,
          "onUpdate:modelValue": r[0] || (r[0] = (p) => n(e)._value = p),
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "editor-style": n(t).style,
          class: g(n(s)),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          placeholder: n(t).placeholder,
          mask: n(t).mask ?? void 0,
          "slot-char": n(t).slotChar ?? "_",
          "auto-clear": n(t).autoClear ?? !0,
          unmask: n(t).unmask ?? !1,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onBlur: i
        }, null, 8, ["id", "modelValue", "disabled", "readonly", "editor-style", "class", "tabindex", "aria-label", "aria-labelledby", "placeholder", "mask", "slot-char", "auto-clear", "unmask", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), Kt = /* @__PURE__ */ b({
  __name: "PrimeInputNumber",
  props: {
    context: Object
  },
  setup(a) {
    const e = a.context, t = e == null ? void 0 : e.attrs;
    function i(r) {
      e == null || e.handlers.blur(r.value);
    }
    function s(r) {
      e == null || e.node.input(r.value);
    }
    const o = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (r, u) => {
      const p = y("InputNumber");
      return f(), c("div", null, [
        _(p, {
          modelValue: n(e)._value,
          "onUpdate:modelValue": u[0] || (u[0] = (m) => n(e)._value = m),
          "input-id": n(e).id,
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "input-style": n(t).style,
          "input-class": n(o),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          placeholder: n(t).placeholder,
          "use-grouping": n(t).useGrouping ?? !0,
          "min-fraction-digits": n(t).minFractionDigits ?? void 0,
          "max-fraction-digits": n(t).maxFractionDigits ?? void 0,
          locale: n(t).locale ?? void 0,
          mode: n(t).mode ?? void 0,
          currency: n(t).currency ?? void 0,
          prefix: n(t).prefix ?? void 0,
          suffix: n(t).suffix ?? void 0,
          "show-buttons": n(t).showButtons ?? void 0,
          "button-layout": n(t).buttonLayout ?? "stacked",
          step: n(t).step ?? void 0,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onInput: s,
          onBlur: i
        }, null, 8, ["modelValue", "input-id", "disabled", "readonly", "input-style", "input-class", "tabindex", "aria-label", "aria-labelledby", "placeholder", "use-grouping", "min-fraction-digits", "max-fraction-digits", "locale", "mode", "currency", "prefix", "suffix", "show-buttons", "button-layout", "step", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), xt = {
  key: 0,
  class: "formkit-prime-left"
}, Zt = {
  key: 1,
  class: "formkit-prime-right"
}, Ht = /* @__PURE__ */ b({
  __name: "PrimeInputSwitch",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      var r;
      e == null || e.node.input((r = l.context) == null ? void 0 : r._value);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      const u = y("InputSwitch");
      return f(), c("div", {
        class: g(n(t).option_class)
      }, [
        n(e).attrs.labelLeft ? (f(), c("span", xt, k(n(e).attrs.labelLeft), 1)) : w("", !0),
        _(u, {
          modelValue: n(e)._value,
          "onUpdate:modelValue": r[0] || (r[0] = (p) => n(e)._value = p),
          "input-id": n(e).id,
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "input-style": n(t).style,
          "input-class": n(s),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          "true-value": n(t).trueValue ?? void 0,
          "false-value": n(t).falseValue ?? void 0,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onInput: i
        }, null, 8, ["modelValue", "input-id", "disabled", "readonly", "input-style", "input-class", "tabindex", "aria-label", "aria-labelledby", "true-value", "false-value", "pt", "pt-options", "unstyled"]),
        n(e).attrs.labelRight ? (f(), c("span", Zt, k(n(e).attrs.labelRight), 1)) : w("", !0)
      ], 2);
    };
  }
}), Wt = /* @__PURE__ */ b({
  __name: "PrimeInputText",
  props: {
    context: Object
  },
  setup(a) {
    const e = a.context, t = e == null ? void 0 : e.attrs;
    function i() {
      return (e == null ? void 0 : e.iconLeft) && (e == null ? void 0 : e.iconLeft.length) > 0;
    }
    function s() {
      return (e == null ? void 0 : e.iconRight) && (e == null ? void 0 : e.iconRight.length) > 0;
    }
    function o() {
      let m = "";
      return i() && (m = `${m}p-input-icon-left `), s() && (m = `${m}p-input-icon-right `), m;
    }
    function r(m) {
      e == null || e.handlers.blur(m.target.value);
    }
    function u(m) {
      e == null || e.node.input(m.target.value);
    }
    const p = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (m, I) => {
      const oe = y("InputText");
      return f(), c("div", null, [
        X("span", {
          class: g(o())
        }, [
          i() ? (f(), c("i", {
            key: 0,
            class: g(n(e).iconLeft)
          }, null, 2)) : w("", !0),
          _(oe, {
            id: n(e).id,
            modelValue: n(e)._value,
            "onUpdate:modelValue": I[0] || (I[0] = (ue) => n(e)._value = ue),
            disabled: n(t)._disabled ?? !1,
            readonly: n(t)._readonly ?? !1,
            style: C(n(t).style),
            class: g(n(p)),
            tabindex: n(t).tabindex,
            "aria-label": n(t).ariaLabel,
            "aria-labelledby": n(t).ariaLabelledby,
            placeholder: n(t).placeholder,
            pt: n(t).pt,
            "pt-options": n(t).ptOptions,
            unstyled: n(t).unstyled ?? !1,
            onInput: u,
            onBlur: r
          }, null, 8, ["id", "modelValue", "disabled", "readonly", "style", "class", "tabindex", "aria-label", "aria-labelledby", "placeholder", "pt", "pt-options", "unstyled"]),
          s ? (f(), c("i", {
            key: 1,
            class: g(n(e).iconRight)
          }, null, 2)) : w("", !0)
        ], 2)
      ]);
    };
  }
}), Yt = /* @__PURE__ */ b({
  __name: "PrimeTextarea",
  props: {
    context: Object
  },
  setup(a) {
    const e = a.context, t = e == null ? void 0 : e.attrs;
    function i(r) {
      e == null || e.handlers.blur(r.target.value);
    }
    function s(r) {
      e == null || e.node.input(r.target.value);
    }
    const o = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (r, u) => {
      const p = y("Textarea");
      return f(), c("div", null, [
        _(p, {
          id: n(e).id,
          modelValue: n(e)._value,
          "onUpdate:modelValue": u[0] || (u[0] = (m) => n(e)._value = m),
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          style: C(n(t).style),
          class: g(n(o)),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          placeholder: n(t).placeholder,
          rows: n(e).rows ?? 3,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onInput: s,
          onBlur: i
        }, null, 8, ["id", "modelValue", "disabled", "readonly", "style", "class", "tabindex", "aria-label", "aria-labelledby", "placeholder", "rows", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), qt = /* @__PURE__ */ b({
  __name: "PrimeKnob",
  props: {
    context: Object
  },
  setup(a) {
    const e = a.context, t = e == null ? void 0 : e.attrs;
    function i(r) {
      e == null || e.node.input(r), e == null || e.handlers.blur(r);
    }
    function s(r) {
      e == null || e.node.input(r);
    }
    const o = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (r, u) => {
      const p = y("Knob");
      return f(), c("div", null, [
        _(p, {
          id: n(e).id,
          modelValue: n(e)._value,
          "onUpdate:modelValue": [
            u[0] || (u[0] = (m) => n(e)._value = m),
            s
          ],
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          style: C(n(t).style),
          class: g(n(o)),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          min: n(t).min ?? 0,
          max: n(t).max ?? 100,
          step: n(t).step ?? void 0,
          size: n(t).size ?? 100,
          "stroke-width": n(t).strokeWidth ?? 14,
          "show-value": n(t).showValue ?? !0,
          "value-color": n(t).valueColor ?? void 0,
          "range-color": n(t).rangeColor ?? void 0,
          "text-color": n(t).textColor ?? void 0,
          "value-template": n(t).valueTemplate ?? void 0,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onChange: i
        }, null, 8, ["id", "modelValue", "disabled", "readonly", "style", "class", "tabindex", "aria-label", "aria-labelledby", "min", "max", "step", "size", "stroke-width", "show-value", "value-color", "range-color", "text-color", "value-template", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), Jt = /* @__PURE__ */ b({
  __name: "PrimeMultiSelect",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      var r;
      e == null || e.node.input((r = l.context) == null ? void 0 : r._value);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      const u = y("MultiSelect");
      return f(), c("div", null, [
        _(u, {
          modelValue: n(e)._value,
          "onUpdate:modelValue": r[0] || (r[0] = (p) => n(e)._value = p),
          "input-id": n(e).id,
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "list-style": n(t).style,
          class: g(n(s)),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          options: n(t).options,
          "option-label": n(t).optionLabel ?? "label",
          "option-value": n(t).optionValue ?? "value",
          filter: n(t).filter ?? !1,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onChange: i
        }, null, 8, ["modelValue", "input-id", "disabled", "readonly", "list-style", "class", "tabindex", "aria-label", "aria-labelledby", "options", "option-label", "option-value", "filter", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), Qt = /* @__PURE__ */ b({
  __name: "PrimeListbox",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      var r;
      e == null || e.node.input((r = l.context) == null ? void 0 : r._value);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      var p;
      const u = y("Listbox");
      return f(), c("div", null, [
        _(u, {
          id: n(e).id,
          modelValue: n(e)._value,
          "onUpdate:modelValue": r[0] || (r[0] = (m) => n(e)._value = m),
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "list-style": n(t).style,
          class: g(n(s)),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          options: (p = n(t)) == null ? void 0 : p.options,
          "option-label": n(t).optionLabel ?? "label",
          "option-value": n(t).optionValue ?? "value",
          multiple: n(t).multiple ?? !1,
          filter: n(t).filter ?? !1,
          "filter-icon": n(t).filterIcon,
          "filter-placeholder": n(t).filterPlaceholder,
          "filter-locale": n(t).filterLocale,
          "filter-match-mode": n(t).filterMatchMode,
          "auto-option-focus": n(t).autoOptionFocus ?? !0,
          "select-on-focus": n(t).selectOnFocus ?? !1,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onChange: i
        }, null, 8, ["id", "modelValue", "disabled", "readonly", "list-style", "class", "tabindex", "aria-label", "aria-labelledby", "options", "option-label", "option-value", "multiple", "filter", "filter-icon", "filter-placeholder", "filter-locale", "filter-match-mode", "auto-option-focus", "select-on-focus", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), Xt = /* @__PURE__ */ b({
  __name: "PrimePassword",
  props: {
    context: Object
  },
  setup(a) {
    const e = a.context, t = e == null ? void 0 : e.attrs;
    function i(r) {
      e == null || e.handlers.blur(r.target.value);
    }
    function s(r) {
      e == null || e.node.input(r.target.value);
    }
    const o = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (r, u) => {
      const p = y("Password");
      return f(), c("div", null, [
        _(p, {
          modelValue: n(e)._value,
          "onUpdate:modelValue": u[0] || (u[0] = (m) => n(e)._value = m),
          "input-id": n(e).id,
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "input-style": n(t).style,
          "input-class": n(o),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          placeholder: n(t).placeholder,
          "medium-regex": n(t).mediumRegex ?? "^(((?=.*[a-z])(?=.*[A-Z]))|((?=.*[a-z])(?=.*[0-9]))|((?=.*[A-Z])(?=.*[0-9])))(?=.{6,})",
          "strong-regex": n(t).strongRegex ?? "^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.{8,})",
          "prompt-label": n(t).promptLabel,
          "weak-label": n(t).weakLabel,
          "medium-label": n(t).mediumLabel,
          "strong-label": n(t).strongLabel,
          "hide-icon": n(t).hideIcon ?? "pi pi-eye-slash",
          "show-icon": n(t).showIcon ?? "pi pi-eye",
          feedback: n(e).feedback ?? !1,
          "toggle-mask": n(e).toggleMask ?? !1,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onInput: s,
          onBlur: i
        }, null, 8, ["modelValue", "input-id", "disabled", "readonly", "input-style", "input-class", "tabindex", "aria-label", "aria-labelledby", "placeholder", "medium-regex", "strong-regex", "prompt-label", "weak-label", "medium-label", "strong-label", "hide-icon", "show-icon", "feedback", "toggle-mask", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), Gt = ["for"], ea = /* @__PURE__ */ b({
  __name: "PrimeRadioButton",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      var r;
      e == null || e.node.input((r = l.context) == null ? void 0 : r._value);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      const u = y("RadioButton");
      return f(), c("div", {
        class: g(n(t).options_class)
      }, [
        (f(!0), c(fe, null, ce(n(t).options, (p) => (f(), c("div", {
          key: p.value,
          class: g(n(t).option_class)
        }, [
          _(u, {
            modelValue: n(e)._value,
            "onUpdate:modelValue": r[0] || (r[0] = (m) => n(e)._value = m),
            name: n(t).name,
            value: p.value,
            "input-style": n(t).style,
            "input-class": n(s),
            pt: n(t).pt,
            "pt-options": n(t).ptOptions,
            unstyled: n(t).unstyled ?? !1,
            onClick: i,
            onChange: i
          }, null, 8, ["modelValue", "name", "value", "input-style", "input-class", "pt", "pt-options", "unstyled"]),
          X("label", {
            for: p.value
          }, k(p.label), 9, Gt)
        ], 2))), 128))
      ], 2);
    };
  }
}), ta = /* @__PURE__ */ b({
  __name: "PrimeRating",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      var r;
      e == null || e.node.input((r = l.context) == null ? void 0 : r._value);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      const u = y("Rating");
      return f(), c("div", null, [
        _(u, {
          id: n(e).id,
          modelValue: n(e)._value,
          "onUpdate:modelValue": r[0] || (r[0] = (p) => n(e)._value = p),
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          style: C(n(t).style),
          class: g(n(s)),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          stars: n(t).stars ?? 5,
          cancel: n(t).cancel ?? !1,
          "on-icon": n(t).onIcon ?? "pi pi-star-fill",
          "off-icon": n(t).offIcon ?? "pi pi-star",
          "cancel-icon": n(t).cancelIcon ?? "pi pi-ban",
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onChange: i
        }, null, 8, ["id", "modelValue", "disabled", "readonly", "style", "class", "tabindex", "aria-label", "aria-labelledby", "stars", "cancel", "on-icon", "off-icon", "cancel-icon", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), aa = /* @__PURE__ */ b({
  __name: "PrimeSlider",
  props: {
    context: Object
  },
  setup(a) {
    const e = a.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      e == null || e.node.input(o), e == null || e.handlers.blur(o);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      const u = y("Slider");
      return f(), c("div", null, [
        _(u, {
          id: n(e).id,
          modelValue: n(e)._value,
          "onUpdate:modelValue": r[0] || (r[0] = (p) => n(e)._value = p),
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          style: C(n(t).style),
          class: g(n(s)),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          min: n(t).min ?? 0,
          max: n(t).max ?? 100,
          step: n(t).step ?? void 0,
          range: n(t).range ?? !1,
          orientation: n(t).orientation ?? "horizontal",
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onChange: i
        }, null, 8, ["id", "modelValue", "disabled", "readonly", "style", "class", "tabindex", "aria-label", "aria-labelledby", "min", "max", "step", "range", "orientation", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), na = /* @__PURE__ */ b({
  __name: "PrimeToggleButton",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      var r;
      e == null || e.node.input((r = l.context) == null ? void 0 : r._value);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      const u = y("ToggleButton");
      return f(), c("div", null, [
        _(u, {
          modelValue: n(e)._value,
          "onUpdate:modelValue": r[0] || (r[0] = (p) => n(e)._value = p),
          "input-id": n(e).id,
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "input-style": n(t).style,
          "input-class": n(s),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          "on-label": n(t).onLabel ?? "Yes",
          "off-label": n(t).offLabel ?? "No",
          "on-icon": n(t).onIcon ?? "pi pi-check",
          "off-icon": n(t).offIcon ?? "pi pi-times",
          "icon-pos": n(t).iconPos ?? "left",
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onChange: i
        }, null, 8, ["modelValue", "input-id", "disabled", "readonly", "input-style", "input-class", "tabindex", "aria-label", "aria-labelledby", "on-label", "off-label", "on-icon", "off-icon", "icon-pos", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), la = /* @__PURE__ */ b({
  __name: "PrimeSelectButton",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      var r;
      e == null || e.node.input((r = l.context) == null ? void 0 : r._value);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      const u = y("SelectButton");
      return f(), c("div", null, [
        _(u, {
          id: n(e).id,
          modelValue: n(e)._value,
          "onUpdate:modelValue": r[0] || (r[0] = (p) => n(e)._value = p),
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          style: C(n(t).style),
          class: g(n(s)),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          options: n(t).options,
          "option-label": n(t).optionLabel ?? "label",
          "option-value": n(t).optionValue ?? "value",
          "option-disabled": n(t).optionDisabled,
          multiple: n(t).multiple ?? !1,
          unselectable: n(t).unselectable ?? !0,
          "data-key": n(t).dataKey,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onChange: i
        }, null, 8, ["id", "modelValue", "disabled", "readonly", "style", "class", "tabindex", "aria-label", "aria-labelledby", "options", "option-label", "option-value", "option-disabled", "multiple", "unselectable", "data-key", "pt", "pt-options", "unstyled"])
      ]);
    };
  }
}), ia = {
  key: 0,
  class: "formkit-prime-left"
}, sa = {
  key: 1,
  class: "formkit-prime-right"
}, ra = /* @__PURE__ */ b({
  __name: "PrimeTriStateCheckbox",
  props: {
    context: Object
  },
  setup(a) {
    const l = a, e = l.context, t = e == null ? void 0 : e.attrs;
    function i(o) {
      var r;
      e == null || e.node.input((r = l.context) == null ? void 0 : r._value);
    }
    const s = h(() => e != null && e.state.validationVisible && !(e != null && e.state.valid) ? `${t == null ? void 0 : t.class} p-invalid` : t == null ? void 0 : t.class);
    return (o, r) => {
      const u = y("TriStateCheckbox");
      return f(), c("div", null, [
        n(e).attrs.labelLeft ? (f(), c("span", ia, k(n(e).attrs.labelLeft), 1)) : w("", !0),
        _(u, {
          modelValue: n(e)._value,
          "onUpdate:modelValue": r[0] || (r[0] = (p) => n(e)._value = p),
          "input-id": n(e).id,
          disabled: n(t)._disabled ?? !1,
          readonly: n(t)._readonly ?? !1,
          "input-style": n(t).style,
          "input-class": n(s),
          tabindex: n(t).tabindex,
          "aria-label": n(t).ariaLabel,
          "aria-labelledby": n(t).ariaLabelledby,
          pt: n(t).pt,
          "pt-options": n(t).ptOptions,
          unstyled: n(t).unstyled ?? !1,
          onClick: i
        }, null, 8, ["modelValue", "input-id", "disabled", "readonly", "input-style", "input-class", "tabindex", "aria-label", "aria-labelledby", "pt", "pt-options", "unstyled"]),
        n(e).attrs.labelRight ? (f(), c("span", sa, k(n(e).attrs.labelRight), 1)) : w("", !0)
      ]);
    };
  }
}), oa = v(Rt, {
  props: []
}), ua = v(Wt, {
  props: ["iconRight", "iconLeft"]
}), pa = v(Kt, {
  props: ["iconRight", "iconLeft"]
}), da = v(Bt, {
  props: []
}), fa = v(Xt, {
  props: ["feedback", "toggleMask"]
}), ca = v(Yt, {
  props: ["rows"]
}), ma = v(Tt, {
  props: []
}), ba = v(Ht, {
  props: []
}), ya = v(Nt, {
  props: []
}), _a = v(Ut, {
  props: []
}), va = v(Jt, {
  props: []
}), ha = v(Qt, {
  props: []
}), ga = v(Mt, {
  props: []
}), $a = v(aa, {
  props: []
}), wa = v(ta, {
  props: []
}), ka = v(ea, {
  props: []
}), Ca = v(Ft, {
  props: []
}), Va = v(qt, {
  props: []
}), La = v(zt, {
  props: []
}), Ia = v(na, {
  props: []
}), Sa = v(la, {
  props: []
}), Oa = v(ra, {
  props: []
}), Da = {
  primeAutoComplete: oa,
  primeInputText: ua,
  primeInputNumber: pa,
  primeInputMask: da,
  primePassword: fa,
  primeCheckbox: ma,
  primeInputSwitch: ba,
  primeTextarea: ca,
  primeEditor: ya,
  primeDropdown: _a,
  primeMultiSelect: va,
  primeCalendar: ga,
  primeSlider: $a,
  primeChips: Ca,
  primeKnob: Va,
  primeRating: wa,
  primeRadioButton: ka,
  primeColorPicker: La,
  primeToggleButton: Ia,
  primeListbox: ha,
  primeSelectButton: Sa,
  primeTriStateCheckbox: Oa
};
export {
  oa as primeAutoCompleteDefinition,
  ga as primeCalendarDefinition,
  ma as primeCheckboxDefinition,
  Ca as primeChipsDefinition,
  La as primeColorPickerDefinition,
  _a as primeDropdownDefinition,
  ya as primeEditorDefinition,
  da as primeInputMaskDefinition,
  pa as primeInputNumberDefinition,
  ba as primeInputSwitchDefinition,
  ua as primeInputTextDefinition,
  Da as primeInputs,
  Va as primeKnobDefinition,
  ha as primeListboxDefinition,
  va as primeMultiSelectDefinition,
  fa as primePasswordDefinition,
  ka as primeRadioButtonDefinition,
  wa as primeRatingDefinition,
  Sa as primeSelectButtonDefinition,
  $a as primeSliderDefinition,
  ca as primeTextareaDefinition,
  Ia as primeToggleButtonDefinition,
  Oa as primeTriStateCheckboxDefinition
};
